bundle/edu.gemini.qpt.client/src/main/java/edu/gemini/qpt/core/listeners/LimitsListener.java

```diff
@@ -66,9 +82,45 @@ public void propertyChange(PropertyChangeEvent evt) {
 			int percentVisible = (int) (100 * a.getMean(Circumstance.TIMING_WINDOW_OPEN, false));
 			switch (percentVisible) {
 			case 100: // good!
-				if (a.getObs().getTimingWindows().size() > 0)
+				if (a.getObs().getTimingWindows().size() > 0 && night != null)
 					mm.addMarker(true, this, Severity.Info, "Timing constraint is met.", v, a);
-				break;
+
+                    // Report timing windows
+                    final TimingWindowSolver timingWindowSolver = new TimingWindowSolver(a.getObs());
+                    final Iterator<Interval> timingWindowIntervals = timingWindowSolver.solve(night).iterator();
+
+                    if (timingWindowIntervals.hasNext()) {
```

Comment:
If you're willing to wander into the seamy underbelly, you might get some mileage out of the `edu.gemini.shared.util.immutable` stuff.

```java
final ImList<Interval> timingWindowIntervals = DefaultImList.create(timingWindowSolver.solve(night));
if (timingWindowIntervals.nonEmpty()) {
    final String msg = timingWindowIntervals.map(i -> formatInterval(i)).mkString("Must be in the timing window ", ", ", ".");
    mm.addMarker(false, this, Severity.Notice, msg, v, a);
}
``` 

assuming you extract the code that does `Interval => String` in a method called `formatInterval`. (See next comment)

Reply:
I already attempted to use `ImOption` but got into a mess with the switch cases, but this approach looks like very easy to implement.